pkg/component/component_processor.go

```diff
@@ -0,0 +1,235 @@
+		var environment string
+		var stage string
+		var tenantFound bool
+		var environmentFound bool
+		var stageFound bool
+
+		for i, part := range stackNamePatternParts {
+			if part == "{tenant}" {
+				tenant = stackParts[i]
+			} else if part == "{environment}" {
+				environment = stackParts[i]
+			} else if part == "{stage}" {
+				stage = stackParts[i]
+			}
+		}
+
+		for stackName := range stacksMap {
+			componentSection, componentVarsSection, _, err = findComponentConfig(stackName, stacksMap, "terraform", component)
+			if err != nil {
+				componentSection, componentVarsSection, _, err = findComponentConfig(stackName, stacksMap, "helmfile", component)
+				if err != nil {
+					continue
+				}
+			}
+
+			tenantFound = true
+			environmentFound = true
+			stageFound = true
+
+			// Search for tenant in stack
+			if len(tenant) > 0 {
+				if tenantInStack, ok := componentVarsSection["tenant"].(string); !ok || tenantInStack != tenant {
+					tenantFound = false
+				}
+			}
+
+			// Search for environment in stack
+			if len(environment) > 0 {
+				if environmentInStack, ok := componentVarsSection["environment"].(string); !ok || environmentInStack != environment {
+					environmentFound = false
+				}
+			}
+
+			// Search for stage in stack
+			if len(stage) > 0 {
+				if stageInStack, ok := componentVarsSection["stage"].(string); !ok || stageInStack != stage {
+					stageFound = false
+				}
+			}
+
+			if tenantFound == true && environmentFound == true && stageFound == true {
+				stack = stackName
+				break
+			}
+		}
+
+		if tenantFound == false || environmentFound == false || stageFound == false {
+			return nil, errors.New(fmt.Sprintf("\nCould not find config for the component '%s' in the stack '%s'.\n"+
+				"Check that all attributes in the stack name pattern '%s' are defined in stack config files.\n"+
+				"Are the component and stack names correct? Did you forget an import?",
+				component,
+				stack,
+				c.Config.Stacks.NamePattern,
+			))
+		}
+	}
+
+	baseComponentName := ""
+	if baseComponent, baseComponentExist := componentSection["component"]; baseComponentExist {
+		baseComponentName = baseComponent.(string)
+	}
+
+	// workspace
+	var workspace string
+	if len(baseComponentName) == 0 {
+		workspace = stack
+	} else {
+		workspace = fmt.Sprintf("%s-%s", stack, component)
+	}
+	componentSection["workspace"] = strings.Replace(workspace, "/", "-", -1)
+
+	return componentSection, nil
+}
+
+// ProcessComponentFromContext accepts context (tenant, environment, stage) and returns the component configuration in the stack
+func ProcessComponentFromContext(component string, tenant string, environment string, stage string) (map[string]interface{}, error) {
+	var stack string
+
+	err := c.InitConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(c.Config.Stacks.NamePattern) < 1 {
+		return nil, errors.New("stack name pattern must be provided in 'stacks.name_pattern' config or 'ATMOS_STACKS_NAME_PATTERN' ENV variable")
```

Comment:
line 50 is the same as this line

It would be nice to have an errors.go module that housed each of these error messages so could call upon them using `errors.stack_pattern_mismatch` or similar to keep it DRY

Reply:
yes, common modules would be nice

Reply:
those are the same since the functions are used at diff times, and they both need `NamePattern` so they check if it exists